base/compiler/abstractinterpretation.jl

```diff
@@ -617,9 +617,13 @@ function abstract_call_method(interp::AbstractInterpreter, method::Method, @nosp
         # this edge is known to terminate
         effects = Effects(effects; terminates=ALWAYS_TRUE)
     elseif edgecycle
-        # Some sort of recursion was detected. Even if we did not limit types,
-        # we cannot guarantee that the call will terminate
-        effects = Effects(effects; terminates=ALWAYS_FALSE)
+        # Some sort of recursion was detected.
+        if edge !== nothing && !edgelimited && !is_edge_recursed(edge, sv)
+            # no `MethodInstance` cycles -- don't taint :terminate
```

Comment:
This seems like it may only consider whether there was a cycle that will impact the computation of the result type. If we note that the result is discarded, there won't be an `edge` or `edgecycle `here at all (note line 512 above).

IIUC, the correct fix for that may be to poison the effects at line 512 however?

Reply:
If we discard the result, we propagated the unknown effects `Effects()`, whose `:terminate` effect is also tainted, and it will eventually poison the rest of the call graph.